tests/apps/qchem/test_dynamics.py

```diff
@@ -94,3 +100,91 @@ def test_evolution(time, unitary, frequency, prob):
         p = eng.run(gbs).state.all_fock_probs()
 
     assert np.allclose(p, prob)
+
+
+@pytest.mark.parametrize("d", [d1, d2])
```

Comment:
These tests are fine, but again, I think they are not detecting whether the sampling is working properly. For example, if we apply a different unitary in the code, it won't get caught by the tests. I suggest coding an example where there is no randomness in the sampling. One example could be to use an identity interferometer, or a unitary that permutes the modes. In both cases, the output state is another Fock state that always gives the same result upon a Fock measurement

Reply:
For me the best test would be one that inspects the program and:
- ensures that the operations are occurring in the expected order
- the unitary and rotation values are as expected

We can then trust SF to sample correctly given that we know we've created the correct program.

So for me, we'd build upon the `test_invalid_loss`/`test_loss` tests and instead have one test to check the order of the operations and another test(s) that looks at the interferometers and rotations.

Regarding the `test_invalid_loss`/`test_loss` tests, you can access the circuit using `call_history[0].circuit`. I then had a quick look at the embedded unitary using `c[3].op.p`. However, I think the `MagicMock` approach used in the `sample.sample` tests should also be a more conventional way of accessing the called arguments.

Reply:
Tests are great now!